### DISCUSSION TITLE
Fail2ban stops detecting postfix-sasl auth failures

### DISCUSSION THREAD

**Discussion by ShuaBlue** [fail2ban/fail2ban#3354] [General]
I have been going nuts trying to figure this out. I am using CentOS 9 stream, with postfix 2:3.5.9-19.el9 and f2b 0.11.2-12.el9. 

I have both postfix and postfix-sasl jails enabled. I get attempts to login to the postfix submission port and they fail with authentication failure. F2b seems to pick up a few of them as "found" right after starting the f2b service. Then, after some time, f2b stops detecting any more failed auth attempts.

Auth login attempt that was caught by f2b:
```Sep 16 22:48:30 postfix postfix/submission/smtpd[2563]: connect from unknown[5.253.204.58]
Sep 16 22:48:44 postfix postfix/submission/smtpd[2563]: Anonymous TLS connection established from unknown[5.253.204.58]: TLSv1.2 with cipher ECDHE-RSA-AES256-GCM-SHA384 (256/256 bits)
Sep 16 22:48:47 postfix postfix/submission/smtpd[2563]: warning: unknown[5.253.204.58]: SASL LOGIN authentication failed: authentication failure
Sep 16 22:48:47 postfix postfix/submission/smtpd[2563]: disconnect from unknown[5.253.204.58] ehlo=2 starttls=1 auth=0/1 quit=1 commands=4/5
```
F2b sees this and detects it:
```2022-09-16 22:48:48,648 fail2ban.filter         [2437]: INFO    [postfix-sasl] Found 5.253.204.58 - 2022-09-16 22:48:47```

Then, some time later, another similar event is shown in postfix log:
```Sep 16 23:22:51 postfix postfix/submission/smtpd[2605]: connect from unknown[94.46.179.80]
Sep 16 23:23:05 postfix postfix/submission/smtpd[2605]: Anonymous TLS connection established from unknown[94.46.179.80]: TLSv1.2 with cipher ECDHE-RSA-AES256-GCM-SHA384 (256/256 bits)
Sep 16 23:23:08 postfix postfix/submission/smtpd[2605]: warning: unknown[94.46.179.80]: SASL LOGIN authentication failed: authentication failure
Sep 16 23:23:08 postfix postfix/submission/smtpd[2605]: disconnect from unknown[94.46.179.80] ehlo=2 starttls=1 auth=0/1 quit=1 commands=4/5
```

regex seems to find the offender
But nothing this time f2b does not log it and doesn't have any other info in the log file. Any subsequent auth failures are not detected from any IP. If I restart f2b service, then the IPs are detected, but eventually they stop being detected again after some time.

This is my jail status:
```Status
- Number of jail:      2
- Jail list:   postfix, postfix-sasl
```

This is the configuration for postfix-sasl:
```
[postfix-sasl]
enabled = true
findtime = 86400
bantime = 604800
maxretry = 3
filter   = postfix[mode=auth]
port     = smtp,465,submission,imap,imaps,pop3,pop3s
logpath  = %(postfix_log)s
backend  = %(postfix_backend)s
```

**Comment by sebres** [comment]
If it does basically work and just sporadically stops at some point, it may look like the backend doesn't get notification about logfile changes. It is good question why.

> If I restart f2b service, then the IPs are detected

Also this is a sign for that. 
If your backend is not systemd (the jail monitoring log-files and not a journal) - can it be that fail2ban doesn't recognize log rotation for some reason or vice versa postfix still continues to write to the rotated log (but created new one which is now monitored by fail2ban).

Which backend the jail uses you can see with `fail2ban-client -d | grep 'add.*postfix-sasl'`.

To check whether the backend has some issue, you can try to switch to alternate one, e. g.:
```diff
  [postfix-sasl]
  # backend = auto
- backend = pyinotify
+ backend = polling
```
(or vice versa)... And then observe whether it doesn't stop to detect failures anymore.

If this is definitely log-rotation (what I guess), you can also force a recreation of log-file in the rotation config or add `fail2ban-client reload postfix-sassl` in postrotate. And then check it with manual rotation.

Alternatively you can configure postfix to log in systemd journal (and switch the backend of jail to `systemd` + restart the jail).

If nothing helps you could try to switch loglevel of fail2ban to debug or heavydebug (but note that it is too wordy for production mode)... 
But you could at least test it with manual rotation.

If fail2ban stops to log completely (and not postfix jail only), maybe this is similar to #2943.

Anyway it is too hard to estimate what it is without additional info (which backend, which logfile, whether it gets rotated or not, rsyslog forwarding log files or not, etc).

**Reply by ShuaBlue** [reply]
Thanks for responding about this issue.


>If your backend is not systemd (the jail monitoring log-files and not a journal) - can it be that fail2ban doesn't recognize log rotation for some reason or vice versa postfix still continues to write to the rotated log (but created new one which is now monitored by fail2ban).

The backend is systemd. This is the output of the command you mentioned:
`['add', 'postfix-sasl', 'systemd']`

>If this is definitely log-rotation (what I guess)

There is log rotation enabled for postfix, but the rotation is setup in a way that the last rotation for /var/log/maillog was on 8/24/2022 (there is not a lot of traffic on this server yet). So the log rotation should not be an issue for recent events.


>If nothing helps you could try to switch loglevel of fail2ban to debug or heavydebug (but note that it is too wordy for production mode)...

I have tried setting this to DEBUG, but nothing comes up in the f2b debug log when there is an auth failure in the postfix log, just as when the log level is set to INFO. I will attempt to set this to HEAVYDEBUG and see if there is any difference.

>If fail2ban stops to log completely (and not postfix jail only), maybe this is similar to https://github.com/fail2ban/fail2ban/issues/2943.

The log just kind of "stalls" at one point and nothing is detected in f2b. However, if I run a `fail2ban-client reload` I definitely see new entries in the log.

**Reply by sebres** [reply]
> the rotation is setup in a way that the last rotation for /var/log/maillog ...

Backend systemd monitors journals and therefore files in /var/log are not interesting at all.

> So the log rotation should not be an issue for recent events.

No, it should not because the monitoring of systemd journal is not affected by the rotation (at least must not).
Fail2ban uses pythonic API for that ([python-systemd/journal](https://github.com/systemd/python-systemd/blob/master/systemd/journal.py)). I'm not aware it may have known issues like that (also not found any issue on GH or even in internet).

> The log just kind of "stalls" at one point and nothing is detected in f2b. 

I guess nothing means also another jails, right? 
If it is not a lack of logging of fail2ban which stops for some reason (as already said something similar to #2943), then this looks much more strange.
One could see whether it only stops to log:
- either trying to ban something manually, for instance with:
  `fail2ban-client set "$some_jail" banip 192.0.2.1; fail2ban-client set "$some_jail" unbanip 192.0.2.1`
  (and then see f2b has logged something)
- and/or write some line to journal so some jail would find it (or do an failed login-attempt) and check `fail2ban-client status "$some_jail"` show increased `Currently failed` or `Total failed`.
- or with `fail2ban-client flushlogs` - this is supposed to force log-flush process, e. g. after log-rotation of f2b-log, and this is normally also a default postrotate: https://github.com/fail2ban/fail2ban/blob/2df58c5281facd73e73778a734a90f7d4fe95468/files/fail2ban-logrotate#L11-L12

If it is not just logging and it really stops to work (monitor journal) I'm still not following why it affecting all jails (are there some other jails with non-systemd backend?)...

>  However, if I run a `fail2ban-client reload` I definitely see new entries in the log.

Weird. The `reload` is just refreshing the runtime configuration of fail2ban, it does not restart jails or somehow affecting monitoring thread by systemd backend (the log-files of file-related backends get refreshed, but not systemd backend).
The only thing happening with systemd backend is flushing all journal matches and adding back hereafter (e. g. `journal.flush_matches()` + `journal.add_matches()` that normally just refresh the pre-filtering inside journal module).
So I'm confused still more now.
